### PR TITLE
Refs #34074 -- Used headers argument for RequestFactory and Client in docs and tests.

### PR DESCRIPTION
**Pull request by David-Wobrock**
Follow-up/clean-up of https://github.com/django/django/pull/16163 ([ticket-34074](https://code.djangoproject.com/ticket/34074))

With help from https://github.com/adamchainz/django-upgrade/pull/305, but also manually fixing.
Ping @adamchainz to see the result the current state of the `django-upgrade` PR through the 3 commits.

We will squash all commits before merging, the 3 commits are for comprehension of the `django-upgrade` work :)

### CODE REVIEW

File: tests/cache/tests.py [hunk 1]
```diff
@@ -2095,9 +2095,9 @@ def test_cache_key_varies_by_url(self):
         """
         get_cache_key keys differ by fully-qualified URL instead of path
         """
-        request1 = self.factory.get(self.path, HTTP_HOST="sub-1.example.com")
+        request1 = self.factory.get(self.path, headers={"host": "sub-1.example.com"})
```
Comment (David-Wobrock) [inline tests/cache/tests.py:2098]: The `django-upgrade` PR is mostly not recognising the `self.factory` nor `self.rf` as being a `RequestFactory` instance.

File: tests/auth_tests/test_views.py [hunk 2]
```diff
@@ -214,7 +214,7 @@ def test_poisoned_http_host(self):
             response = self.client.post(
                 "/password_reset/",
                 {"email": "staffmember@example.com"},
-                headers={"host": "www.example:dr.frankenstein@evil.tld"}
+                headers={"host": "www.example:dr.frankenstein@evil.tld"},
```
Comment (David-Wobrock) [inline tests/auth_tests/test_views.py:217]: Only `black` changes were applied during the pre-commit hooks.

**Comment by felixxm** [comment]
@David-Wobrock Is it ready for review?

**Comment by David-Wobrock** [comment]
> @David-Wobrock Is it ready for review?

@felixxm I'd say it is, or at least a good start :)

**Comment by felixxm** [comment]
@David-Wobrock Thanks :+1: 

I found a release blocker while checking this patch, see ticket-34240.

**Pull request merged**